File: yagooglesearch/searcher.py
package yagooglesearchpackage

import (
	"encoding/json"
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const version = "1.10.0"

var (
	rootLogger  *log.Logger
	userAgents  []string
	resultLangs []string
	userAgent   = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36"
)

func init() {
	// Initialize logging
	logFile, err := os.OpenFile("yagooglesearch.go.log", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("Error opening log file: %v", err)
	}
	rootLogger = log.New(logFile, "", log.Ldate|log.Ltime|log.Lshortfile)

	// Load user agents
	installFolder := filepath.Dir(os.Args[0])
	userAgentsFile := filepath.Join(installFolder, "user_agents.txt")
	if _, err := os.Stat(userAgentsFile); err == nil {
		data, err := ioutil.ReadFile(userAgentsFile)
		if err != nil {
			rootLogger.Fatalf("Error reading user_agents.txt: %v", err)
		}
		userAgents = strings.Split(string(data), "\n")
	} else {
		userAgents = []string{userAgent}
	}

	// Load result languages
	resultLangsFile := filepath.Join(installFolder, "result_languages.txt")
	if _, err := os.Stat(resultLangsFile); err == nil {
		data, err := ioutil.ReadFile(resultLangsFile)
		if err != nil {
			rootLogger.Fatalf("Error reading result_languages.txt: %v", err)
		}
		lines := strings.Split(string(data), "\n")
		for _, line := range lines {
			if parts := strings.SplitN(line, "=", 2); len(parts) == 2 {
				resultLangs = append(resultLangs, parts[0])
			}
		}
	}
}

func getTbs(fromDate, toDate time.Time) string {
	return fmt.Sprintf("cdr:1,cd_min:%s,cd_max:%s", fromDate.Format("01/02/2006"), toDate.Format("01/02/2006"))
}

type SearchClient struct {
	query                           string
	tld                             string
	langHtmlUI                      string
	langResult                      string
	tbs                             string
	safe                            string
	start                           int
	num                             int
	country                         string
	extraParams                     map[string]string
	maxSearchResultURLsToReturn     int
	minimumDelayBetweenPagedResults int
	userAgent                       string
	manageHttp429s                  bool
	http429CoolOffTime              int
	http429CoolOffFactor            float64
	proxy                           string
	verifySSL                       bool
	verbosity                       int
	verboseOutput                   bool
	googleExemption                 string
	cookies                         map[string]string
	urlParameters                   []string
}

func NewSearchClient(
	query string,
	tld string,
	langHtmlUI string,
	langResult string,
	tbs string,
	safe string,
	start int,
	num int,
	country string,
	extraParams map[string]string,
	maxSearchResultURLsToReturn int,
	minimumDelayBetweenPagedResults int,
	userAgent string,
	manageHttp429s bool,
	http429CoolOffTime int,
	http429CoolOffFactor float64,
	proxy string,
	verifySSL bool,
	verbosity int,
	verboseOutput bool,
	googleExemption string,
) *SearchClient {
	sc := &SearchClient{
		query:                           query,
		tld:                             tld,
		langHtmlUI:                      langHtmlUI,
		langResult:                      langResult,
		tbs:                             tbs,
		safe:                            safe,
		start:                           start,
		num:                             num,
		country:                         country,
		extraParams:                     extraParams,
		maxSearchResultURLsToReturn:     maxSearchResultURLsToReturn,
		minimumDelayBetweenPagedResults: minimumDelayBetweenPagedResults,
		userAgent:                       userAgent,
		manageHttp429s:                  manageHttp429s,
		http429CoolOffTime:              http429CoolOffTime,
		http429CoolOffFactor:            http429CoolOffFactor,
		proxy:                           proxy,
		verifySSL:                       verifySSL,
		verbosity:                       verbosity,
		verboseOutput:                   verboseOutput,
		googleExemption:                 googleExemption,
		urlParameters:                   []string{"btnG", "cr", "hl", "num", "q", "safe", "start", "tbs", "lr"},
	}

	if sc.userAgent == "" {
		sc.userAgent = sc.assignRandomUserAgent()
	}

	if sc.langResult != "" && !contains(resultLangs, sc.langResult) {
		rootLogger.Printf("%s is not a valid language result. Setting lang_result to \"lang_en\".", sc.langResult)
		sc.langResult = "lang_en"
	}

	if sc.num > 100 {
		rootLogger.Println("The largest value allowed by Google for num is 100. Setting num to 100.")
		sc.num = 100
	}

	if sc.maxSearchResultURLsToReturn > 400 {
		rootLogger.Println("yagooglesearch is usually only able to retrieve a maximum of ~400 results.")
	}

	if sc.googleExemption != "" {
		sc.cookies = map[string]string{"GOOGLE_ABUSE_EXEMPTION": sc.googleExemption}
	}

	return sc
}

func (sc *SearchClient) assignRandomUserAgent() string {
	return userAgents[rand.Intn(len(userAgents))]
}

func (sc *SearchClient) filterSearchResultURLs(link string) string {
	if strings.HasPrefix(link, "/url?") || strings.HasPrefix(link, "http://www.google.com/url?") {
		parsedURL, err := url.Parse(link)
		if err != nil {
			return ""
		}
		queryParams := parsedURL.Query()
		if q, ok := queryParams["q"]; ok {
			link = q[0]
		} else if u, ok := queryParams["url"]; ok {
			link = u[0]
		}
	}

	parsedURL, err := url.Parse(link)
	if err != nil || parsedURL.Host == "" || strings.Contains(strings.ToLower(parsedURL.Host), "google") {
		return ""
	}

	return link
}

func (sc *SearchClient) http429Detected() {
	sc.http429CoolOffTime = int(float64(sc.http429CoolOffTime) * sc.http429CoolOffFactor)
	rootLogger.Printf("Increasing HTTP 429 cool off time to %d minutes", sc.http429CoolOffTime)
}

func (sc *SearchClient) getPage(url1 string) (string, error) {
	client := &http.Client{}
	if sc.proxy != "" {
		proxyURL, err := url.Parse(sc.proxy)
		if err != nil {
			return "", err
		}
		client.Transport = &http.Transport{Proxy: http.ProxyURL(proxyURL)}
	}

	req, err := http.NewRequest("GET", url1, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", sc.userAgent)
	if sc.cookies != nil {
		for key, value := range sc.cookies {
			req.AddCookie(&http.Cookie{Name: key, Value: value})
		}
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode == 429 {
		if !sc.manageHttp429s {
			return "HTTP_429_DETECTED", nil
		}
		rootLogger.Printf("Sleeping for %d minutes...", sc.http429CoolOffTime)
		time.Sleep(time.Duration(sc.http429CoolOffTime) * time.Minute)
		sc.http429Detected()
		return sc.getPage(url1)
	}

	return string(body), nil
}

func (sc *SearchClient) search() ([]map[string]string, error) {
	var results []map[string]string
	totalValidLinksFound := 0

	if sc.extraParams == nil {
		sc.extraParams = make(map[string]string)
	}

	for builtinParam := range sc.extraParams {
		if contains(sc.urlParameters, builtinParam) {
			return nil, fmt.Errorf("GET parameter \"%s\" is overlapping with the built-in GET parameter", builtinParam)
		}
	}

	_, err := sc.getPage(fmt.Sprintf("https://www.google.%s/", sc.tld))
	if err != nil {
		return nil, err
	}

	for totalValidLinksFound < sc.maxSearchResultURLsToReturn {
		rootLogger.Printf("Stats: start=%d, num=%d, total_valid_links_found=%d / max_search_result_urls_to_return=%d",
			sc.start, sc.num, totalValidLinksFound, sc.maxSearchResultURLsToReturn)

		var url string
		if sc.start > 0 {
			if sc.num == 10 {
				url = fmt.Sprintf("https://www.google.%s/search?hl=%s&lr=%s&q=%s&start=%d&tbs=%s&safe=%s&cr=%s&filter=0",
					sc.tld, sc.langHtmlUI, sc.langResult, sc.query, sc.start, sc.tbs, sc.safe, sc.country)
			} else {
				url = fmt.Sprintf("https://www.google.%s/search?hl=%s&lr=%s&q=%s&start=%d&num=%d&tbs=%s&safe=%s&cr=%s&filter=0",
					sc.tld, sc.langHtmlUI, sc.langResult, sc.query, sc.start, sc.num, sc.tbs, sc.safe, sc.country)
			}
		} else {
			if sc.num == 10 {
				url = fmt.Sprintf("https://www.google.%s/search?hl=%s&lr=%s&q=%s&tbs=%s&safe=%s&cr=%s&filter=0",
					sc.tld, sc.langHtmlUI, sc.langResult, sc.query, sc.tbs, sc.safe, sc.country)
			} else {
				url = fmt.Sprintf("https://www.google.%s/search?hl=%s&lr=%s&q=%s&num=%d&tbs=%s&safe=%s&cr=%s&filter=0",
					sc.tld, sc.langHtmlUI, sc.langResult, sc.query, sc.num, sc.tbs, sc.safe, sc.country)
			}
		}

		for key, value := range sc.extraParams {
			url += fmt.Sprintf("&%s=%s", key, value)
		}

		body, err := sc.getPage(url)
		if err != nil {
			return nil, err
		}

		if body == "HTTP_429_DETECTED" {
			results = append(results, map[string]string{"error": "HTTP_429_DETECTED"})
			return results, nil
		}

		doc, err := goquery.NewDocumentFromReader(strings.NewReader(body))
		if err != nil {
			return nil, err
		}

		validLinksFoundInThisSearch := 0
		doc.Find("a").Each(func(_ int, s *goquery.Selection) {
			link, exists := s.Attr("href")
			if !exists {
				return
			}

			link = sc.filterSearchResultURLs(link)
			if link == "" {
				return
			}

			if sc.verboseOutput {
				title := s.Text()
				description := s.Parent().Parent().Contents().Eq(1).Text()
				if description == "" {
					description = s.Parent().Parent().Contents().Eq(2).Text()
				}

				results = append(results, map[string]string{
					"rank":        fmt.Sprintf("%d", totalValidLinksFound+1),
					"title":       title,
					"description": description,
					"url":         link,
				})
			} else {
				results = append(results, map[string]string{"url": link})
			}

			totalValidLinksFound++
			validLinksFoundInThisSearch++
			if totalValidLinksFound >= sc.maxSearchResultURLsToReturn {
				return
			}
		})

		if validLinksFoundInThisSearch == 0 {
			rootLogger.Println("No valid search results found on this page. Moving on...")
			break
		}

		sc.start += sc.num
		time.Sleep(time.Duration(rand.Intn(12)+sc.minimumDelayBetweenPagedResults) * time.Second)
	}

	return results, nil
}

func contains(slice []string, item string) bool {
	for _, a := range slice {
		if a == item {
			return true
		}
	}
	return false
}

func main() {
	// Example usage
	sc := NewSearchClient(
		"example query",
		"com",
		"en",
		"lang_en",
		"",
		"off",
		0,
		10,
		"",
		nil,
		100,
		7,
		"",
		true,
		60,
		1.1,
		"",
		true,
		5,
		false,
		"",
	)

	results, err := sc.search()
	if err != nil {
		rootLogger.Fatalf("Error during search: %v", err)
	}

	jsonResults, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		rootLogger.Fatalf("Error marshaling results to JSON: %v", err)
	}

	fmt.Println(string(jsonResults))
}

File: yagooglesearch/searcher.go
package yagooglesearch

import (
	"fmt"
	"github.com/gocolly/colly"
	"github.com/gocolly/colly/proxy"
	"github.com/gocolly/colly/queue"
	"log"
	"net/url"
	"strings"
	"sync"
	"time"
)

// SearchClient 结构体保持不变
type SearchClient struct {
	Config           SearchClientConfig
	Collector        *colly.Collector
	Queue            *queue.Queue
	searchResultList []any
	foundUrls        map[string]bool
	wg               sync.WaitGroup // 【新】使用 WaitGroup 来进行同步
}

// SearchClientConfig 接口保持不变
type SearchClientConfig struct {
	Query                         string
	Tld                           string
	LangHtmlUI                    string
	LangResult                    string
	Num                           int
	Start                         int
	MaxSearchResultUrlsToReturn   int
	UserAgent                     string
	Proxies                       []string
	Verbosity                     bool
	Parallelism                   int
	RandomDelay                   time.Duration
	YagooglesearchManagesHttp429s bool
	Http429CoolOffTimeInMinutes   int
}

// NewSearchClient 接口保持不变
func NewSearchClient(config SearchClientConfig) (*SearchClient, error) {
	// (这部分基本不变，只是初始化了 SearchClient)
	if config.Query == "" {
		return nil, fmt.Errorf("query cannot be empty")
	}
	if config.Tld == "" {
		config.Tld = "com"
	}
	if config.LangHtmlUI == "" {
		config.LangHtmlUI = "en"
	}
	if config.Num == 0 {
		config.Num = 100
	}
	if config.Parallelism == 0 {
		config.Parallelism = 1
	} // 【注意】默认并行度改为1，增加初始成功率
	if config.RandomDelay == 0 {
		config.RandomDelay = 5 * time.Second
	} // 默认增加延迟

	c := colly.NewCollector(
		colly.Async(true),
		colly.UserAgent(config.UserAgent),
		// 【新】自动处理Cookie
		colly.AllowURLRevisit(),
	)

	// 【新】增强请求头，使其更像浏览器
	c.OnRequest(func(r *colly.Request) {
		r.Headers.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9")
		r.Headers.Set("Accept-Language", "en-US,en;q=0.9")
		if r.Ctx.Get("is_search") != "" {
			r.Headers.Set("Referer", "https://www.google.com/")
		}
		if config.Verbosity {
			log.Println("Visiting", r.URL.String())
		}
	})

	c.Limit(&colly.LimitRule{
		DomainGlob:  "*.google.*",
		Parallelism: config.Parallelism,
		RandomDelay: config.RandomDelay,
	})

	if len(config.Proxies) > 0 {
		rp, err := proxy.RoundRobinProxySwitcher(config.Proxies...)
		if err != nil {
			return nil, fmt.Errorf("failed to create proxy switcher: %w", err)
		}
		c.SetProxyFunc(rp)
	}

	q, _ := queue.New(config.Parallelism, &queue.InMemoryQueueStorage{MaxSize: 10000})

	sc := &SearchClient{
		Config:           config,
		Collector:        c,
		Queue:            q,
		searchResultList: []any{},
		foundUrls:        make(map[string]bool),
	}

	// 统一设置回调
	sc.setupCallbacks()
	return sc, nil
}

// setupCallbacks 包含所有的事件处理逻辑
func (sc *SearchClient) setupCallbacks() {
	c := sc.Collector

	// 【核心改造】只使用colly内置方法来解析HTML
	c.OnHTML("div#search div.g", func(e *colly.HTMLElement) {
		// e 现在代表一个 <div class="g"> 元素，也就是一个搜索结果块

		if len(sc.searchResultList) >= sc.Config.MaxSearchResultUrlsToReturn {
			return
		}

		// 1. 查找链接URL
		// 我们需要找这个div.g块里面的第一个<a>标签的href属性
		// 注意：ChildAttr找到的是第一个匹配的子元素，这通常就是标题链接
		link := e.ChildAttr("a", "href")
		// (可以加入你之前的 filterSearchResultUrls 逻辑)
		// ...

		// 2. 过滤和去重
		if link != "" && strings.HasPrefix(link, "http") && !strings.Contains(link, "google.com") {
			if _, found := sc.foundUrls[link]; !found {
				sc.foundUrls[link] = true

				// 3. 如果需要详细输出，提取标题和描述
				if sc.Config.Verbosity {
					// 使用 e.ChildText 查找标题(h3)和描述
					// e 是 <div class="g">, h3 是它的子元素
					title := e.ChildText("h3")

					// Google的描述通常在一个特定的div里，我们也可以用ChildText，
					// 但选择器可能需要更精确。这个选择器表示在<div class="g">下
					// 找一个同样是div且属性data-sncf='2'的元素
					desc := e.ChildText("div[data-sncf='2']")

					result := map[string]any{
						"rank":        len(sc.searchResultList) + 1,
						"title":       strings.TrimSpace(title),
						"description": strings.TrimSpace(desc),
						"url":         link,
					}
					sc.searchResultList = append(sc.searchResultList, result)
				} else {
					// 只保存URL
					sc.searchResultList = append(sc.searchResultList, link)
				}

				if sc.Config.Verbosity {
					log.Printf("Found unique URL #%d: %s", len(sc.searchResultList), link)
				}
			}
		}
	})

	// OnHTML: 查找“下一页”链接
	c.OnHTML("a#pnnext[href]", func(e *colly.HTMLElement) {
		if len(sc.searchResultList) >= sc.Config.MaxSearchResultUrlsToReturn {
			log.Println("Max search results reached. Not visiting next page.")
			return
		}
		// 【关键】找到了下一页，就把它加入队列
		nextPageLink := e.Request.AbsoluteURL(e.Attr("href"))
		log.Println("Found next page link, adding to queue:", nextPageLink)
		sc.wg.Add(1) // 增加一个等待任务
		sc.Queue.AddURL(nextPageLink)
	})

	// OnResponse: 检查是否是验证页面，并在所有请求完成后通知WaitGroup
	c.OnResponse(func(r *colly.Response) {
		// 检查是否是被拦截的页面
		bodyStr := string(r.Body)
		if strings.Contains(bodyStr, "如果您在几秒钟内没有被重定向") || strings.Contains(bodyStr, "enablejs") {
			log.Println("!!! Google human verification page detected! Try using a proxy or reducing parallelism/delay. !!!")
		}
	})

	// OnError: 处理错误
	c.OnError(func(r *colly.Response, err error) {
		log.Printf("Request URL: %s failed with response code %d, Error: %v", r.Request.URL, r.StatusCode, err)
		sc.wg.Done() // 【重要】请求出错，也必须算作任务完成
	})

	// OnScraped: 当一个页面的所有OnHTML回调都执行完毕后触发
	c.OnScraped(func(r *colly.Response) {
		// 只有在搜索页面完成后，才标记任务完成
		if r.Ctx.Get("is_search") != "" {
			log.Printf("Finished scraping: %s", r.Request.URL)
			sc.wg.Done() // 【重要】一个搜索任务完成
		}
	})
}

// 【核心改造】Search 函数
func (sc *SearchClient) Search() ([]any, error) {
	// 【阶段一：预热】
	// 先访问一次首页，让colly的cookie jar获取初始cookie，这个过程是同步的。
	log.Println("Phase 1: Warming up session by visiting Google homepage...")
	warmupCollector := sc.Collector.Clone() // 使用克隆的collector进行预热，避免干扰主任务
	err := warmupCollector.Visit(fmt.Sprintf("https://www.google.%s/", sc.Config.Tld))
	if err != nil {
		log.Printf("Warning: Failed to warm up session, but will proceed. Error: %v", err)
	} else {
		log.Println("Warm-up successful, session cookies should be stored.")
	}

	// 【阶段二：执行搜索】
	log.Println("Phase 2: Starting main search task...")

	// 构造初始搜索URL
	baseURL := fmt.Sprintf("https://www.google.%s/search", sc.Config.Tld)
	params := url.Values{}
	params.Set("q", sc.Config.Query)
	params.Set("num", fmt.Sprintf("%d", sc.Config.Num))
	params.Set("hl", sc.Config.LangHtmlUI)
	initialURL := baseURL + "?" + params.Encode()

	// 设置上下文，标记这是一个搜索请求
	ctx := colly.NewContext()
	ctx.Put("is_search", "true")

	// 启动队列，但现在我们用WaitGroup来等待
	//go func() {
	// q.Run会阻塞，直到队列为空并且没有活动的消费者
	// 这意味着当所有翻页任务都完成后，它会自动退出
	sc.Queue.Run(sc.Collector)
	//}()

	// 将第一个任务加入队列，并设置等待数为1
	sc.wg.Add(1)
	err = sc.Queue.AddURL(initialURL)
	if err != nil {
		sc.wg.Done() // 如果添加失败，也得减少等待数
		return nil, fmt.Errorf("failed to add initial URL: %w", err)
	}

	// 【关键】等待所有任务（初始搜索+所有翻页）完成
	sc.wg.Wait()

	log.Println("Search process finished.")
	return sc.searchResultList, nil
}
